votor/src/event_handler.rs

```diff
@@ -127,50 +142,72 @@ impl EventHandler {
         }
 
         while !exit.load(Ordering::Relaxed) {
+            let mut receive_event_time = Measure::start("receive_event");
             let event = select! {
                 recv(event_receiver) -> msg => {
                     msg?
                 },
                 default(Duration::from_secs(1))  => continue
             };
+            receive_event_time.stop();
+            local_context.stats.receive_event_time = local_context
+                .stats
+                .receive_event_time
+                .saturating_add(receive_event_time.as_us() as u32);
 
             if event.should_ignore(vctx.root_bank_cache.root_bank().slot()) {
+                local_context.stats.ignored = local_context.stats.ignored.saturating_add(1);
                 continue;
             }
 
+            let mut event_processing_time = Measure::start("event_processing");
+            let event_index = EventHandlerStats::event_to_index(&event);
             let votes = Self::handle_event(
-                &mut my_pubkey,
                 event,
                 &timer_manager,
                 &ctx,
                 &mut vctx,
                 &rctx,
-                &mut pending_blocks,
-                &mut finalized_blocks,
-                &mut received_shred,
+                &mut local_context,
             )?;
+            event_processing_time.stop();
+            local_context
+                .stats
+                .incr_event_with_timing(event_index, event_processing_time.as_us());
 
-            // TODO: properly bubble up error handling here and in call graph
+            let mut send_vote_time = Measure::start("send_vote");
             for vote in votes {
+                local_context.stats.incr_vote(&vote);
                 vctx.bls_sender.send(vote).map_err(|_| SendError(()))?;
             }
+            send_vote_time.stop();
+            local_context.stats.send_vote_time = local_context
+                .stats
+                .send_vote_time
+                .saturating_add(send_vote_time.as_us() as u32);
+            local_context.stats.maybe_report();
         }
 
         Ok(())
     }
 
     fn handle_event(
-        my_pubkey: &mut Pubkey,
         event: VotorEvent,
         timer_manager: &RwLock<TimerManager>,
         ctx: &SharedContext,
         vctx: &mut VotingContext,
         rctx: &RootContext,
-        pending_blocks: &mut PendingBlocks,
-        finalized_blocks: &mut BTreeSet<Block>,
-        received_shred: &mut BTreeSet<Slot>,
+        local_context: &mut LocalContext,
     ) -> Result<Vec<BLSOp>, EventLoopError> {
         let mut votes = vec![];
+        let LocalContext {
+            ref mut my_pubkey,
+            ref mut pending_blocks,
+            ref mut finalized_blocks,
+            ref mut received_shred,
+            ref mut stats,
+        .. // The `..` is a good practice to ignore other fields you don't need
```

Comment:
think this is common enough to not need a comment

Reply:
Removed.